akka-stream-tests/src/test/scala/akka/stream/io/InputStreamSourceSpec.scala

```diff
@@ -4,54 +4,110 @@
+
+import scala.util.Success
 import scala.concurrent.duration._
 
 class InputStreamSourceSpec extends StreamSpec(UnboundedMailboxConfig) {
 
   val settings = ActorMaterializerSettings(system).withDispatcher("akka.actor.default-dispatcher")
   implicit val materializer = ActorMaterializer(settings)
 
-  "InputStreamSource" must {
+  private def inputStreamFor(bytes: Array[Byte]): InputStream =
+    new ByteArrayInputStream(bytes)
 
-    "not signal when no demand" in {
-      val f = StreamConverters.fromInputStream(() =>
-        new InputStream {
-          override def read(): Int = 42
-        })
+  "InputStream Source" must {
 
-      Await.result(f.takeWithin(5.seconds).runForeach(it => ()), 10.seconds)
+    "read bytes from InputStream" in assertAllStagesStopped {
+      val f =
+        StreamConverters.fromInputStream(() => inputStreamFor(Array('a', 'b', 'c').map(_.toByte))).runWith(Sink.head)
+
+      f.futureValue should ===(ByteString("abc"))
     }
 
-    "read bytes from InputStream" in assertAllStagesStopped {
-      val f = StreamConverters
+    "record number of bytes read" in assertAllStagesStopped {
+      StreamConverters
+        .fromInputStream(() => inputStreamFor(Array(1, 2, 3)))
+        .toMat(Sink.ignore)(Keep.left)
+        .run
+        .futureValue shouldEqual IOResult(3, Success(Done))
+    }
+
+    "return fail if close fails" in assertAllStagesStopped {
+      val fail = new RuntimeException("oh dear")
+      StreamConverters
         .fromInputStream(() =>
           new InputStream {
-            @volatile var buf = List("a", "b", "c").map(_.charAt(0).toInt)
-            override def read(): Int = {
-              buf match {
-                case head :: tail =>
-                  buf = tail
-                  head
-                case Nil =>
-                  -1
-              }
-
-            }
+            override def read(): Int = -1
+            override def close(): Unit = throw fail
           })
-        .runWith(Sink.head)
+        .toMat(Sink.ignore)(Keep.left)
+        .run
+        .failed
+        .futureValue shouldEqual IOOperationIncompleteException(0, fail)
+    }
 
-      f.futureValue should ===(ByteString("abc"))
+    "return failure if creation fails" in {
+      val fail = new RuntimeException("oh dear indeed")
+      StreamConverters
+        .fromInputStream(() => {
+          throw fail
+        })
+        .toMat(Sink.ignore)(Keep.left)
+        .run
+        .failed
+        .futureValue shouldEqual IOOperationIncompleteException(0, fail)
+    }
+
+    "handle failure on read" in assertAllStagesStopped {
+      val fail = new RuntimeException("oh dear indeed")
+      StreamConverters
+        .fromInputStream(() => () => throw fail)
+        .toMat(Sink.ignore)(Keep.left)
+        .run
+        .failed
+        .futureValue shouldEqual IOOperationIncompleteException(0, fail)
+    }
+
+    "fail materialized value if downstream doesn't read all of it" in {
```

Comment:
I'm not sure this is the desired behavior: it can be perfectly valid for downstream not to read everything, right? 

The 'old' `postStop` also did `completionPromise.trySuccess(IOResult(readBytesTotal, Success(Done)))` regardless of whether the stream was completely consumed or not.